File: frontend/src/order/object/core/OrderLineItem.ts
import type {
  CurrencyCents,
  DisplayText,
  MenuItemId,
  Quantity,
} from '@/shared/object/domain/DomainObjects'
import type { OrderItemSelection } from '@/order/object/core/OrderItemSelection'

export type OrderLineItem = {
  menuItemId: MenuItemId
  name: DisplayText
  quantity: Quantity
  unitPriceCents: CurrencyCents
  refundedQuantity: Quantity
  selections: OrderItemSelection[]
}

File: backend/src/main/scala/order/object/core/OrderLineItem.scala
package domain.order

import domain.shared.given

import io.circe.{Decoder, Encoder}
import io.circe.generic.semiauto.*
import domain.shared.*

final case class OrderLineItem(
    menuItemId: MenuItemId,
    name: DisplayText,
    quantity: Quantity,
    unitPriceCents: CurrencyCents,
    refundedQuantity: Quantity,
)
object OrderLineItem:
  given Encoder[OrderLineItem] = deriveEncoder
  given Decoder[OrderLineItem] = Decoder.instance { cursor =>
    for
      menuItemId <- cursor.get[MenuItemId]("menuItemId")
      name <- cursor.get[DisplayText]("name")
      quantity <- cursor.get[Quantity]("quantity")
      unitPriceCents <- cursor.get[CurrencyCents]("unitPriceCents")
      refundedQuantity <- cursor.getOrElse[Quantity]("refundedQuantity")(NumericDefaults.ZeroQuantity)
    yield OrderLineItem(
      menuItemId = menuItemId,
      name = name,
      quantity = quantity,
      unitPriceCents = unitPriceCents,
      refundedQuantity = refundedQuantity,
    )
  }
